### PR TITLE
Assignment2 - Josh Pedersen

### PR DESCRIPTION
**Pull request by Joshua-Pedersen**
Completed assignment

### CODE REVIEW

**Comment by Joshua-Pedersen** [comment]
Unsure why build test is failing to "Post comment on PR"

**Comment by BenjaminMichaelis** [comment]
I see the action is failing on this pr, looking into this now

**Comment by BenjaminMichaelis** [comment]
It's because the GITHUB_TOKEN Permissions in the Set up Job section (line 16 of the results of the workflow) currently has `PullRequests: read` instead of `PullRequests: write` . Not sure yet why...

**Comment by Joshua-Pedersen** [comment]
Seems it still ran as ```read```

**Comment by BenjaminMichaelis** [comment]
Working now, disabled the feature for now. Thank you for your patience; GitHub Permissions are messy sometimes :)

**Comment by Joshua-Pedersen** [comment]
> Working now, disabled the feature for now. Thank you for your patience; GitHub Permissions are messy sometimes :)

Thanks

File: Logger/Logger.csproj [hunk 1]
```diff
@@ -2,7 +2,12 @@
 
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
```
Comment (Joshua-Pedersen) [inline Logger/Logger.csproj:4]: Need to update

**Comment by NathanMarsee** [comment]
I'll review this

File: Logger/LogFactory.cs [hunk 2]
```diff
@@ -2,9 +2,22 @@
...
-
-        return null;
+        if (string.IsNullOrWhiteSpace(_file))
+        {
+            return null;
+        }
+        else
+        {
+            return new FileLogger(_file!)
+            { ClassName = className};
```
Comment (NathanMarsee) [inline Logger/LogFactory.cs:16]: CreateLogger() method is meant to use ConfigureFileLogger() method according to the assignment checklist.

File: Logger/BaseLogger.cs [hunk 3]
```diff
@@ -3,5 +3,7 @@
 public abstract class BaseLogger
 {
     public abstract void Log(LogLevel logLevel, string message);
+    public string? ClassName
+        { get; set; }
```
Comment (NathanMarsee) [inline Logger/BaseLogger.cs:7]: You can just use nameof() when identifying class names to the logger (as listed in the assignment checklist) without needing a variable.

**Review by NathanMarsee — APPROVED** [review]
Assignment Checklist:

- There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
- Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
  - The current date/time ✔
  - The name of the class that created the logger ✔
  - The log level ✔
  - The message ✔
  - The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
- The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ❌
- If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
- Inside of BaseLoggerMixins implement extension methods on BaseLogger for
  - Error, ✔
  - Warning, ✔
  - Information, and ✔
  - Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
- Use the nameof() operator when identifying the class name to the logger ❌
- Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
- Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
- Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
- Ensure that you turn on EnforceCodeStyleInBuild ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌✔
- Turn on Nullability (Nullable) ✔
- All of the above should be unit tested. ✔


Additional notes:
Great work overall - there's just a couple of small ways you can better fulfill the assignment requirements in a few places, but basically everything is here.

File: Logger/LogFactory.cs (hunk 2 above)
Comment (Joshua-Pedersen) [reply Logger/LogFactory.cs:16]: I may be mistaken but I took ```This should take in a file path and store it in a **private member**. It should use this when instantiating a new `FileLogger` in its `CreateLogger` method.``` as a reference to ```CreateLogger``` using the private member ```_file``` within which ```ConfigureFileLogger``` modifies. I'll prob has about this tomorrow though.

File: Logger.Tests/Logger.Tests.csproj [hunk 4]
```diff
@@ -1,9 +1,14 @@
...
+		<TargetFramework>net8.0</TargetFramework>
 		<IsPackable>false</IsPackable>
-		<LangVersion>11.0</LangVersion>
+		<LangVersion>12.0</LangVersion>
+    <Nullable>enable</Nullable>
+    <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
+    <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <EnableNETAnalyzers>true</EnableNETAnalyzers>
 	</PropertyGroup>
```
Comment (BenjaminMichaelis) [inline Logger.Tests/Logger.Tests.csproj:12]:
```suggestion
		<LangVersion>12.0</LangVersion>
		<Nullable>enable</Nullable>
		<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
		<CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
		<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
		<EnableNETAnalyzers>true</EnableNETAnalyzers>
	</PropertyGroup>
```

File: Logger/BaseLogger.cs (hunk 3 above)
Comment (BenjaminMichaelis) [inline Logger/BaseLogger.cs:7]:
```suggestion
    public string ClassName { get; set; }
```
Getter and setter can be on same line preferably in this case.
Why is this nullable?

File: Logger/BaseLoggerMixins.cs [hunk 5]
```diff
@@ -1,6 +1,38 @@
...
 public static class BaseLoggerMixins
 {
+    public static void LoggerNullCheck(BaseLogger? logger)
+    {
+        ArgumentNullException.ThrowIfNull(logger);
+    }
+    
+    public static void Error(this BaseLogger? logger, string message, params object[] args)
+    {
+        LoggerNullCheck(logger);
```
Comment (BenjaminMichaelis) [inline Logger/BaseLoggerMixins.cs:15]: I'm not sure how this method is saving any code in this case. I like the idea, but just seems unnecessary

File: Logger/BaseLoggerMixins.cs [hunk 6]
```diff
@@ -1,6 +1,38 @@
...
+namespace Logger;
 
 public static class BaseLoggerMixins
 {
+    public static void LoggerNullCheck(BaseLogger? logger)
+    {
+        ArgumentNullException.ThrowIfNull(logger);
+    }
+    
+    public static void Error(this BaseLogger? logger, string message, params object[] args)
```
Comment (BenjaminMichaelis) [inline Logger/BaseLoggerMixins.cs:13]: Why is logger nullable? I expect to call this message, a logger is needed.

File: Logger/LogFactory.cs (hunk 2 above)
Comment (BenjaminMichaelis) [reply Logger/LogFactory.cs:16]: `CreateLogger` does not need to call `ConfigureFileLogger`, while nonetheless `ConfigureFileLogger` should be called at somepoint before `CreateLogger`

**Pull request closed without merging**